pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
+       'Na2O', 'K2O', 'P2O5']
+
+    element_AW = {
+        'F': 18.9984032,
+        'Cl': 35.4527,
+        'S': 32.066,
+        'Ni': 58.6934,
+        'Co': 58.93320,
+        'Sr': 87.62,
+        'Ba': 137.327,
+        'Rb': 85.4678,
+        'Cs': 132.90545,
+        'Li': 6.941,
+        'Zr': 91.224,
+        'Cr': 51.9961,
+        'V': 50.9415,
+        'O': 15.9994
+    }
+
+    element_oxide = {
+        'F': 'F',
+        'Cl': 'Cl',
+        'S': 'S',
+        'Ni': 'NiO',
+        'Co': 'CoO',
+        'Sr': 'SrO',
+        'Ba': 'BaO',
+        'Rb': 'Rb2O',
+        'Cs': 'Cs2O',
+        'Li': 'Li2O',
+        'Zr': 'ZrO2',
+        'Cr': 'Cr2O3',
+        'V': 'V2O3'
+    }
+
+    mineral_molecular_weights = {
+        'Q': 60.0843,
+        'Z': 183.3031,
+        'Ks': 154.2803,
+        'An': 278.207276,
+        'Ns': 122.0632,
+        'Ac': 462.0083,
+        'Di': 225.99234699428553,
+        'Tn': 196.0625,
+        'Hy': 109.82864699428553,
+        'Ab': 524.446,
+        'Or': 556.6631,
+        'Wo': 116.1637,
+        'Ol': 159.57299398857106,
+        'Pf': 135.9782,
+        'Ne': 284.1088,
+        'Lc': 436.4945,
+        'Cs': 172.2431,
+        'Kp': 316.3259,
+        'Ap': 328.8691887,
+        'Fr': 94.0762,
+        'Pr': 135.96640000000002,
+        'Cm': 223.83659999999998,
+        'Il': 151.7452,
+        'Cc': 100.0892,
+        'C': 101.9613,
+        'Ru': 79.8988,
+        'Mt': 231.53860000000003,
+        'Hm': 159.6922,
+        'Mg-Di': 216.5504,
+        'Mg-Hy': 100.3887,
+        'Mg-Ol': 140.6931
+    }
+
+
+    # Convert element ppm to Oxides wt%
+
+    trace = ['F', 'Cl', 'S', 'Ni', 'Co',
+               'Sr', 'Ba', 'Rb', 'Cs', 'Li', 'Zr', 'Cr', 'V']
+
+    trace_oxides = ['CO2', 'SO3', 'F', 'Cl', 'S', 'NiO', 'CoO',
+                    'SrO', 'BaO', 'Rb2O', 'Cs2O', 'Li2O', 'ZrO2', 'Cr2O3', 'V2O3']
+
+    def calculate_oxide_1(oxide_weight, element_AW, ppm):
+        return (oxide_weight/element_AW) * ppm * (10**-4)
+
+
+    def calculate_oxide_2(oxide_weight, element_AW, ppm):
+        return (oxide_weight/(2 * element_AW)) * ppm * (10**-4)
+
+
+    for element in trace:
+        if element in ['Ni', 'Co', 'Sr', 'Ba', 'Zr']:
+            oxide_name = element_oxide[element]
+            df[oxide_name] = calculate_oxide_1(
+                                         pt.formula(oxide_name).mass,
+                                         element_AW[element],
+                                         df[element])
+        elif element in ['S', 'Cl', 'F']:
+            df[element] = df[element] * (10**-4)
+        else:
+            oxide_name = element_oxide[element]
+            df[oxide_name] = calculate_oxide_2(
+                                         pt.formula(oxide_name).mass,
+                                         element_AW[element],
+                                         df[element])
+
+
+    # Adjust majors wt% to 100% then adjust again to account for trace components
```

Comment:
I agree with your thinking, however in the Verma paper for whatever reason, it says to renormalise the majors, then renormalise again with the minor/trace. That said it may be worth you checking as I may be reading this wrong.

Reply:
It does seem the case that the paper did this double normalization, but I'm not sure it's clear how normalization adjusted the major components in this calculation (i.e. whether the minor components are taken as given and the scaling for major components is `(100 - sumtraces) / 100` or both are adjusted with `100 / (100 + sumtraces)`)! Leave it in for now, especially if it's giving consistent results.